ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+      MockServer mockServer = mockServers.get(i);
+      if (mockServer.getDataCenter().equals(dcWherePutHappened)) {
+        mockServer.setServerErrorForAllRequests(ServerErrorCode.No_Error);
+        break;
+      }
+    }
+    getAndAssertSuccess();
+  }
+
+  // @todo: test the case where read is not called before chunks come in.
+  // @todo: test the case where read comes in immediately after callback, and chunks come in delayed.
+  // @todo: test the case where operation callback is called with exception.
+  // @todo: test the case where async write results in an exception. - read should be notified,
+  //        operation should get completed.
+  // @todo: test the case where a subsequent chunk fails at the router. the read should fail in that case.
+  // @todo: maybe a test for legacy blobs.
+  // @todo: possibly tests where intermediate chunks get expired/deleted.
+
+  /**
+   * Construct GetBlob operations with appropriate callbacks, then poll those operations until they complete,
+   * and ensure that the whole blob data is read out and the contents match.
+   */
+  private void getAndAssertSuccess()
+      throws Exception {
+    final CountDownLatch readCompleteLatch = new CountDownLatch(1);
+    final AtomicReference<Exception> readCompleteException = new AtomicReference<>(null);
+    final AtomicLong readCompleteResult = new AtomicLong(0);
+    final AtomicReference<Exception> operationException = new AtomicReference<>(null);
+
+    Callback<ReadableStreamChannel> callback = new Callback<ReadableStreamChannel>() {
+      @Override
+      public void onCompletion(final ReadableStreamChannel result, final Exception exception) {
+        if (exception != null) {
+          operationException.set(exception);
+          readCompleteLatch.countDown();
+        } else {
+          Utils.newThread(new Runnable() {
+            @Override
+            public void run() {
+              assertSuccess(result, readCompleteLatch, readCompleteResult, readCompleteException);
+            }
+          }, false).start();
+        }
+      }
+    };
+
+    operationsCount.incrementAndGet();
+    GetBlobOperation op =
+        new GetBlobOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            callback, operationCompleteCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+      }
+    }
+
+    readCompleteLatch.await();
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertNull("Operation exception should be null, was:" + operationException.get(), operationException.get());
```

Comment:
why not this?
```
if(operationException.get() != null) {
 throw operationException.get();
}
```

In the current way, you lose the stacktrace (same thing with `readCompleteException`).